### PR TITLE
Fixes for failing nag tests

### PR DESCRIPTION
**Pull request by slevis-lmwg**
### Description of changes

Fix problems uncovered by adding the -nan compilation flag for the Nag compiler.

### Specific notes

Contributors other than yourself, if any:
@billsacks @ekluzek 

CTSM Issues Fixed (include github issue #):
#1994

Are answers expected to change (and if so in what way)?
Unknown

### CODE REVIEW

File: src/main/surfrdUtilsMod.F90 [hunk 1]
```diff
@@ -58,12 +58,15 @@ subroutine check_sums_equal_1(arr, lb, name, caller, ier, sumto)
...
-    if ( present(sumto) ) TotalSum = sumto
     if( present(ier) ) ier = 0
     found = .false.
 
-    do nl = lbound(arr, 1), ub
+    do nl = lb, ub
+       if ( present(sumto) ) then
+          TotalSum(nl) = sumto(nl)
+       else
+          TotalSum(nl) = 1._r8
```
Comment (slevis-lmwg) [inline src/main/surfrdUtilsMod.F90:68]: As previously written, the code sometimes populated TotalSum with NaNs.

My changes get past this issue, but a new one has surfaced, so the pursuit continues.

**Comment by slevis-lmwg** [comment]
This test `SMS_Ld5_D_P48x1.f10_f10_mg37.IHistClm51Bgc.izumi_nag.clm-decStart` with the -nan compilation flag reveals that
`photo_temp_acclim_timescale = nan` in `subroutine InitTimeAveragingGlobals` in `FatesInterfaceMod.F90`.

- Is this triggered because we are running a non-fates simulation?
- Should a non-fates simulation exercise this code at all?

So running a fates test `SMS_D_Ld5.f45_f45_mg37.I2000Clm51Fates.izumi_nag.clm-FatesCold` but this fails in CanopyFluxes:
`t_ref2m(p) = thm(p) + temp1(p)*dth(p)*(1._r8/temp12m(p) - 1._r8/temp1(p))`
I haven't tested this one with a write statement, yet.

**Comment by slevis-lmwg** [comment]
> This test `SMS_Ld5_D_P48x1.f10_f10_mg37.IHistClm51Bgc.izumi_nag.clm-decStart` with the -nan compilation flag reveals that `photo_temp_acclim_timescale = nan` in `subroutine InitTimeAveragingGlobals` in `FatesInterfaceMod.F90`.

I will try @adrifoster recommendation of putting
`if (use_fates)`
around this call
https://github.com/ESCOMP/CTSM/blob/4a868c1ccd700e78b5a2c2f064441a1d612c57bf/src/main/clm_initializeMod.F90#L340

File: src/main/clm_initializeMod.F90 [hunk 2]
```diff
@@ -337,7 +337,7 @@ subroutine initialize2(ni,nj)
     end if
 
     ! Pass model timestep info to FATES
-    call CLMFatesTimesteps()
+    if (use_fates) call CLMFatesTimesteps()
```
Comment (slevis-lmwg) [inline src/main/clm_initializeMod.F90:340]: Without the `if`, the code populated fates params with NaNs.

This change gets me past this issue, but a new one has surfaced.

**Comment by slevis-lmwg** [comment]
For the latest commit, I ran the izumi test-suite and confirmed that I had not changed answers in tests that pass (i.e. non-nag tests).

Now nag tests show three sources of error depending on which test I look at:
- line 3997: Error occurred in PHOTOSYNTHESISMOD:HYBRID_PHS
- line 1441: Error occurred in CANOPYFLUXESMOD:CANOPYFLUXES
- line 491: Error occurred in EDLOGGINGMORTALITYMOD:GET_HARVESTABLE_CARBON

File: src/main/surfrdUtilsMod.F90 (hunk 1 above)
Comment (slevis-lmwg) [reply src/main/surfrdUtilsMod.F90:68]: From today's stand-up conversation:
@billsacks suggested returning lines 60-69 to how they were.

File: src/biogeochem/CNVegNitrogenStateType.F90 [hunk 3]
```diff
@@ -1157,8 +1157,8 @@ subroutine Summary_nitrogenstate(this, bounds, num_allc, filter_allc, &
          this%totn_patch(bounds%begp:bounds%endp), &
          this%totn_p2c_col(bounds%begc:bounds%endc))
 
-    do fc = 1,num_allc
-       c = filter_allc(fc)
+    do fc = 1,num_soilc
+       c = filter_soilc(fc)
```
Comment (slevis-lmwg) [inline src/biogeochem/CNVegNitrogenStateType.F90:1161]: @billsacks suggested initializing the variable that was giving nans to zero before call p2c. I had tried this last week (at least for the CNVegCarbonStateType error) and failed for some reason, so for now I will keep the do-loop changes. The izumi test-suite suggests that nothing breaks due the change to the do-loops

**Comment by slevis-lmwg** [comment]
For the latest commit, the izumi test-suite confirms that I have not changed answers in tests that pass (i.e. non-nag tests).

Now nag tests show the following sources of error depending on which test I look at:

* line 1929: Error occurred in PHOTOSYNTHESISMOD:PHOTOSYNTHESIS
e.g. SMS_D_Ld5.f10_f10_mg37.I1850Clm45BgcCrop.izumi_nag.clm-crop
* line 1441: Error occurred in CANOPYFLUXESMOD:CANOPYFLUXES
e.g. the fates and non-fates tests listed [above.](https://github.com/ESCOMP/CTSM/pull/2051#issuecomment-1631647285) BUT these tests do not write out the offending variable(s) when I place write statements before line 1441. I have tried the write statements surrounded and not surrounded with `if isnan` and if denominators equal zero.
* line 491: Error occurred in EDLOGGINGMORTALITYMOD:GET_HARVESTABLE_CARBON
e.g. ERS_D_Mmpi-serial_Ld5_Vmct.1x1_brazil.I2000Clm50FatesRs.izumi_nag.clm-FatesCold
ERS_D_Mmpi-serial_Ld5.1x1_brazil.I2000Clm50FatesRs.izumi_nag.clm-FatesCold
The latter returned NaN for variable `site_area` in line 491

**Comment by slevis-lmwg** [comment]
I will commit/push additional updates from `intent(out)` to `intent(inout)`. The izumi test-suite still confirms that I have not changed answers in tests that were already passing.

Now nag tests show the following sources of error:
- line 598: Error occurred in ACCUMULMOD:UPDATE_ACCUM_FIELD_TIMEAVG
e.g. test SMS_D_Ld1_Mmpi-serial.f45_f45_mg37.I2000Clm50SpRs.izumi_nag.clm-ptsRLA
- line 1011: Error occurred in BALANCECHECKMOD:BALANCECHECK
e.g. ERP_D_Ld5_P48x1.f10_f10_mg37.I2000Clm50Sp.izumi_nag.clm-o3lombardozzi2015
- line 1019: Error occurred in SUBGRIDAVEMOD:C2G_1D
e.g. SMS_Ld5_D_P48x1.f10_f10_mg37.IHistClm51Bgc.izumi_nag.clm-decStart
- line 491: Error occurred in EDLOGGINGMORTALITYMOD:GET_HARVESTABLE_CARBON
e.g. ERS_D_Mmpi-serial_Ld5_Vmct.1x1_brazil.I2000Clm50FatesRs.izumi_nag.clm-FatesCold
ERS_D_Mmpi-serial_Ld5.1x1_brazil.I2000Clm50FatesRs.izumi_nag.clm-FatesCold
The latter returned NaN for variable site_area in line 491

**Comment by slevis-lmwg** [comment]
Added tag "next" from conversation with @ekluzek 
- so that we may discuss (at Thurs. morning meeting) whether to bring some or all of the code changes into main before I complete this PR
- Otherwise, this PR is low priority until all the CESM components decide that it becomes high priority

**Comment by slevis-lmwg** [comment]
At a recent Thursday morning software meeting we decided to merge this PR because the changes to-date are good. Nevertheless, I will leave #1994 open while nag tests continue to fail with the -nan compilation flag.

Test-suites comparing to dev137 IN PROGRESS

**Comment by slevis-lmwg** [comment]
Test-suites comparing to dev137:
izumi PASS
cheyenne IN PROGRESS

File: src/biogeophys/FrictionVelocityMod.F90 [hunk 4]
```diff
@@ -750,10 +750,10 @@ subroutine FrictionVelocity(this, lbn, ubn, fn, filtern, &
     real(r8) , intent(in)    :: ur      ( lbn: )         ! wind speed at reference height [m/s] [lbn:ubn]
     real(r8) , intent(in)    :: um      ( lbn: )         ! wind speed including the stablity effect [m/s] [lbn:ubn]
     real(r8) , intent(out)   :: ustar   ( lbn: )         ! friction velocity [m/s] [lbn:ubn]
-    real(r8) , intent(out)   :: temp1   ( lbn: )         ! relation for potential temperature profile [lbn:ubn]
-    real(r8) , intent(out)   :: temp12m ( lbn: )         ! relation for potential temperature profile applied at 2-m [lbn:ubn]
-    real(r8) , intent(out)   :: temp2   ( lbn: )         ! relation for specific humidity profile [lbn:ubn]
-    real(r8) , intent(out)   :: temp22m ( lbn: )         ! relation for specific humidity profile applied at 2-m [lbn:ubn]
+    real(r8) , intent(inout) :: temp1   ( lbn: )         ! relation for potential temperature profile [lbn:ubn] (inout instead of out to prevent returning nan)
```
Comment (ekluzek) [inline src/biogeophys/FrictionVelocityMod.F90:753]: Remove the inout part in the comment in favor of the actual FORTRAN keyword, no need to also have it in the comment.

**Review by ekluzek — APPROVED** [review]
Some small suggestions on comments. Good to have this in, as it's a nice robustness improvment.

**Pull request merged**